cli/flox/src/commands/publish.rs

```diff
@@ -101,21 +118,47 @@ impl Publish {
         let build_metadata =
             check_build_metadata(&flox, &env_metadata, &path_env, &FloxBuildMk, &package)?;
 
-        let cache = if no_store {
+        let cache = if metadata_only {
             None
         } else {
-            merge_cache_options(config.flox.publish, cache_args)?
+            // Get the ingress URI for this catalog if it has one configured.
+            let ingress_uri = flox.catalog_client.get_ingress_uri(&catalog_name).await?;
+            let catalog_has_ingress_uri = ingress_uri.is_some();
+
+            // Determine whether a signing key was supplied.
+            let no_key_in_config = config
+                .flox
+                .publish
+                .as_ref()
+                .and_then(|p_conf| p_conf.signing_key.as_ref())
+                .is_none();
+            let no_key_arg = cache_args.signing_private_key.is_none();
+            let no_key_supplied = no_key_in_config && no_key_arg;
+
+            // It is an error (for now) if a user has a catalog configured that accepts
+            // uploads, but the user has attempted to publish without a signing key,
+            // and has not explicitly asked to only upload metadata.
+            if catalog_has_ingress_uri && no_key_supplied && !metadata_only {
+                let msg = formatdoc! { "
+                   A signing key is required to upload artifacts.
+
+                   You can supply a signing key by either:
+                   - Providing a path to a key with the '--signing-private-key' option.
+                   - Setting it in the config via 'flox config --set publish.signing-key <path>'
+
+                   Or you can publish without uploading artifacts via the '--metadata-only' option.
+                "};
+                bail!(msg);
+            }
+
+            merge_cache_options(config.flox.publish, cache_args, ingress_uri)?
```

Comment:
(follow-up) Duplicating the merge and error (after https://github.com/flox/flox/pull/2872) handling logic between `handle()` and `merge_cache_options()` makes this a bit harder to reason about and would be worth reconsolidating in one place.